csrc/host_ir/host_ir.h

```diff
@@ -114,6 +114,48 @@ class PostOnStream : public Expr {
   }
 };
 
+class StreamIr : public Val {
+ public:
+  StreamIr(IrBuilderPasskey passkey);
+  StreamIr(const StreamIr* src, IrCloner* ir_cloner);
+  bool sameAs(const Statement* other) const override;
+
+  NVFUSER_DECLARE_CLONE
+  std::string toString(int indent_size = 0) const override;
+  std::string toInlineString(int indent_size = 0) const override;
+
+ private:
+  // the class contains a unique ID counter accross instances that is mainly
+  // useful for debug print
+  const int idx_;
+  static int running_counter_;
+};
+
+class SetCurrentStream : public Expr {
```

Comment:
How about `PostOnStream` taking `StreamIr` directly instead of requiring a `SetCurrentStream` to set up some state? Can you help me navigate the trade-offs?

Reply:
This is a good point. Imo the trade-off is as follow
- Advantage of having a separate Ir `SetCurrentStream` instead of using `PostOnStream`:
    - makes the code semantically more explicit and readable
- Disadvantage:
    - increase complexity by having an extra Ir. This new Ir `SetCurrentStream` will probably have no other use case

Given this, I think your suggestion is relevant so let me apply it here. Lmk what you think

Reply:
> Closer to the CUDA program that would be generated

I realized this is debatable: for communication, the generated program will use PG instead of nccl. I failed to find a PG API to take in a pre-created stream, which seems to be a general problem if we want host IR to control stream assignments. Any ideas?

Reply:
> Indeed, b) maps directly to pytorch API as well as FusionExcutor(Cache) API, which makes it very convenient for the time being. 

Makes sense -- you are right in that we currently rely on wrapper APIs (e.g. PG for launching communication, at::empty for allocation, and FusionExecutor::runFusion for launching computation kernels) that take the stream from the context. 

I agree that it makes more sense to keep the SetCurrentStream IR as your previous version did.